doc/DHCPv6_relay/DHCPv6-relay-agent-High-Level-Design.md

```diff
@@ -0,0 +1,215 @@
+
+![image](https://user-images.githubusercontent.com/42761586/117859842-5fd13b00-b244-11eb-9297-c2674d128dd9.png)
+
+# Requirements
+
+- Configured and running DHCPv6 client and server
+- Connectivity between the relay agent and DHCPv6 server
+- Configure one or more IP helper addresses for specified VLANs to forward DHCPv6 requests to DHCPv6 servers on other subnets.
+- IPv6 multicast address for all DHCP agents: IPv6 ff02::1:2
+- Client UDP port:546
+- Server and Relay Agent UDP port: 547
+
+# Topology
+
+![image](https://user-images.githubusercontent.com/42761586/117859824-58aa2d00-b244-11eb-991f-7ad8759f6612.png)
+
+# Design
+
+# CLI and Usage
+
+-show dhcprelay counters
+
+-clear dhcprelay counters
+
+-enable/Disable option 79
+
+-enable/Disable use-loopback-address (for dual tor)
+
+-show/config ip helpers
+
+# DHCPRELAY counter
+
+Keeps count of all relay Messages:
+SOLICIT
+ADVERTISE
+REQUEST
+CONFIRM
+RENEW
+REBIND
+REPLY
+RELEASE
+DECLINE
+RELAY-FORWARD
+RELAY-REPLY
+
+# CONFIG DB schema
+
+DHCP|intf-i|dhcpv6\_servers: [&quot;dhcp-server-0&quot;, &quot;dhcp-server-1&quot;, ...., &quot;dhcp-server-n-1&quot;]
+
+DHCP|intf-i|dhcpv6\_options: []
+
+# Option 79 for client link-layer address
+
+Option 79 should be enabled by default and can be disabled through command line.
+
+# Option for Dual ToR
+
+Relayed DHCPv6 packet from ToR may have the response routed to the peer ToR that has the link as standby. Since the originating client is not active on this ToR, the peer ToR won&#39;t be able to relay the response. Peer ToR will not receive the packets as the originating client is not active on this ToR. Instead of using Vlan SVI IP address, relay agent source address needs to be set to listen on the loopback address. When DHCP server responses are received by relay agent on the peer ToR, DHCP relay agent would then forward the packet to the peer ToR using its loopback IP interface.
+
+# Feature table
+
+Adding to existing DHCP relay container. No new feature table added
+
+# RADV modification
+
+Router sends an Router Advertisement message that indicates to nodes on the network that they should use DHCPv6 as their method of dynamic address configuration. RA message contains A, M, O, L bits. The routers can use two flags in RA messages to tell the attached end hosts which method to use:
+
+- Managed-Config-Flag(M-bit) tells the end-host to use DHCPv6 exclusively;
+- Other-Config-Flag(O-bit) tells the end-host to use SLAAC to get IPv6 address and DHCPv6 to get other parameters such as DNS server address.
+- Absence of both flags tells the end-host to use only SLAAC.
+
+# CoPP manager
+
+Control Plane Policing manager is currently configured to only trap DHCPv6 packets when DHCPv6 is enabled.
+
+# Source IP
+
+VLAN SVI IP
+
+Configurable option to use loopback address for dual ToR
+
+# Multiple VLAN
+
+Check kernel routing table and find mapping between interface and server
```

Comment:
the `link_address` field should help select the appropriate vlan interface. No need to check the kernel routing here.